Prevent concurrent operator execution by removing leases discovery

timoreimann opened the pull request:
With Cilium's minimum supported Kubernetes version being 1.19 and support for leases in Kubernetes having been around since 1.14, we can drop the discovery logic to check for the availability of leases. Instead, we will assume that the capability always exists.

The assumption helps with enforcing HA mode in cilium-operator at times when the API server is temporarily unavailable, preventing the case where multiple HA operator may run in leader mode concurrently (and break networking) due to the existing fallback logic that is now removed.

Fixes: #17870

Continuation of #18122

/cc @aanm @christarazi 

```release-note
Remove Kubernetes leases discovery to enforce HA mode in cilium-operator when the API server is unavailable
```

timoreimann commented on pkg/k8s/version/version.go:
```diff
@@ -271,39 +212,36 @@ func updateK8sServerVersion(client kubernetes.Interface) error {
-			}
-
-			return leasesFallbackDiscovery(client, apiDiscoveryEnabled)
+	if !apiDiscoveryEnabled {
```
Note to reviewers: I reversed the logic (handing `!apiDiscoveryEnabled` first instead of `apiDiscoveryEnabled`) so that I would not need an `else` case and save on one indention level.

bimmlerd reviewed (approved):
agent changes LGTM, but have not thought through the implications for operator

tommyp1ckles commented on pkg/k8s/version/version.go:
```diff
@@ -271,39 +212,36 @@ func updateK8sServerVersion(client kubernetes.Interface) error {
 	err := updateK8sServerVersion(client)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to update Kubernetes server version: %s", err)
```
Should use the `%w` directive for the error

tommyp1ckles commented on pkg/k8s/version/version.go:
```diff
@@ -271,39 +212,36 @@ func updateK8sServerVersion(client kubernetes.Interface) error {
+	// readiness of regular pods which require Cilium to function
+	// correctly. By treating failure to discover API groups as
+	// fatal, a critical loop can be entered in which Cilium cannot
+	// start because the API groups can't be discovered and th API
```
typo: "and th API"

tommyp1ckles replied:
Just realized this is existing code, but we might as well clean these up if we're touching these 😄

tommyp1ckles commented on pkg/k8s/version/version.go:
```diff
@@ -271,39 +212,36 @@ func updateK8sServerVersion(client kubernetes.Interface) error {
+	if err != nil {
+		// It doesn't make sense to retry the retrieval of this
+		// information at a later point because the capabilities are
+		// primiarly used while the agent is starting up. Instead, fall
```
primiarly -> primarily

derailed commented on pkg/k8s/version/version.go:
```diff
@@ -271,39 +212,36 @@ func updateK8sServerVersion(client kubernetes.Interface) error {
-			return leasesFallbackDiscovery(client, apiDiscoveryEnabled)
+	if !apiDiscoveryEnabled {
+		if err := endpointSlicesFallbackDiscovery(client); err != nil {
+			return fmt.Errorf("failed to fallback-discover endpoint slices: %s", err)
```
nit: "%w"

derailed commented on pkg/k8s/version/version.go:
```diff
@@ -271,39 +212,36 @@ func updateK8sServerVersion(client kubernetes.Interface) error {
+		log.WithError(err).Warning("Unable to discover API groups and resources")
 		if err := endpointSlicesFallbackDiscovery(client); err != nil {
-			return err
+			return fmt.Errorf("failed to fallback-discover endpoint slices after API groups discovery failure: %s", err)
```
nit: "%w"

derailed reviewed (commented):
@timoreimann Nice work!

tommyp1ckles reviewed (approved):
Thanks @timoreimann! these changes look good to me.

timoreimann committed:
Prevent concurrent operator execution by removing leases discovery

With Cilium's minimum supported Kubernetes version being 1.19 and
support for leases in Kubernetes having been around since 1.14, we can
drop the discovery logic to check for the availability of leases.
Instead, we will assume that the capability always exists.

The assumption helps with enforcing HA mode in cilium-operator at times
when the API server is temporarily unavailable, preventing the case
where multiple HA operator may run in leader mode concurrently (and
break networking) due to the existing fallback logic that is now
removed.

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

timoreimann committed:
DEBUG: do not use tparse

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

timoreimann committed:
DEBUG: turn off tparse globally

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

timoreimann committed:
Use clientset API for lease lock

Signed-off-by: Timo Reimann <ttr314@googlemail.com>

derailed reviewed (approved):
@timoreimann Thanks for the updates!
